'flutter start' calls 'flutter apk' if necessary.

mpcomplete opened the pull request:
flutter start no longer depends on a pre-built SkyShell.apk. It builds a
new one, as long as an AndroidManifest.xml exists.

We rebuild the .apk every time either AndroidManifest.xml or
flutter.yaml changes.

mpcomplete commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -179,227 +189,338 @@ class ApkCommand extends FlutterCommand {
         help: 'Password for the entry within the keystore.');
   }
 
-  Future _findServices(_ApkComponents components) async {
```
This looks like a lot of changes, but these deletions are all just moves. I moved these methods out of the ApkCommand class to be toplevel functions, so we could call them without an instance of that class.

eseidelGoogle commented:
@devoncarew has been doing most of the flutter_tools changes as of late.

devoncarew commented on packages/flutter_tools/lib/src/commands/start.dart:
```diff
@@ -109,10 +117,12 @@ class StartCommand extends StartCommandBase {
 }
 
 Future<int> startApp(
+  BuildConfiguration config,
```
Can you move this to after `toolchain`? Where we need these args, we generally use `devices`, `applicationPackages`, and `toolchain` in that order.

(though we should look into ways to not have to pass around quite so many variables)

mpcomplete replied:
Done. I also changed it to pass the List of BuildConfigurations, since this command is platform agnostic.

devoncarew commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -179,227 +191,343 @@ class ApkCommand extends FlutterCommand {
-    );
+  _ApkComponents components = await _findApkComponents(config, enginePath, manifest, resources);
+  if (components == null) {
+    printError('Unable to build APK.');
```
Perhaps print why?

devoncarew commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -132,6 +136,14 @@ class _ApkComponents {
+  String password;
+  String keyAlias;
+  String keyPassword;
+  ApkKeystoreInfo({this.keystore, this.password, this.keyAlias, this.keyPassword});
```
nit: Ian is going to want spaces inside the braces

devoncarew commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -179,227 +191,343 @@ class ApkCommand extends FlutterCommand {
+    ];
+  } else {
+    androidSdkPath = AndroidDevice.getAndroidSdkPath();
+    if (androidSdkPath == null) {
```
nit: no need for braces here

devoncarew commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -179,227 +191,343 @@ class ApkCommand extends FlutterCommand {
+}
+
+int _signApk(_ApkBuilder builder, _ApkComponents components, File apk,
+             ApkKeystoreInfo keystoreInfo) {
```
here (and elsewhere), this paran should be on the start of the next line

mpcomplete replied:
Not sure what you mean. Is there an example somewhere? Like this?

int _signApk
(_ApkBuilder builder, _ApkComponents components, File apk, ApkKeystoreInfo keystoreInfo) {
...
}

devoncarew replied:
I meant to say the end paran - thuswise:

```
int _signApk(
  _ApkBuilder builder, _ApkComponents components, File apk, ApkKeystoreInfo keystoreInfo
) {
```

There's one or two examples in here: https://github.com/flutter/engine/blob/master/sky/specs/style-guide.md.

devoncarew commented:
@mpcomplete, could you add sample output from `flutter start`, when there are and aren't changes (i.e., when a build is necessary and when it's not)?

mpcomplete commented:
Build is up-to-date:
>  flutter start
Updating flutter tool...
APK up to date. Skipping build step.
Starting lib/main.dart on Nexus 5...

Build is stale:
> touch flutter.yaml ; flutter start
Updating flutter tool...
Signing the APK using the debug keystore
APK generated: build/app.apk
Installing app.apk on device.
Starting lib/main.dart on Nexus 5...

(Note there's a long pause before the 'Signing' message when rebuilding. Maybe I should add a Status-level log message when rebuilding the APK)

devoncarew commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -179,227 +191,342 @@ class ApkCommand extends FlutterCommand {
 
-    builder.sign(keystore, keystorePassword, keyAlias, keyPassword, apk);
+// Returns true if the apk is out of date and needs to be rebuilt.
+bool _needsRebuild(String apkPath, String manifest) {
```
Perhaps also include `.packages` in this check? That will capture dependencies changes, including an engine update.

devoncarew commented:
Looks like there's a few more lint issues in `apk.dart`, related to string interpolation.

Here's a bunch of thoughts, in no particular order:
- when there's no building required, `start` could just say `Starting foo on bar...` - we don't need to say `APK up to date. Skipping build step.` We should still say something about skipping the build if the user has explicitly asked for a build like via `flutter apk` (or perhaps never do timestamp checks for explicit builds?)
- agree w/ that long pause at the start, that we should say something. For things we know will take a while, perhaps also add the duration at the end. So `Building APK...`, and then `APK built to build/app.apk (5.3s).`
- what can the user currently do if they know that the app needs to be re-built, but our system thinks it's up-to-date? Is there a `build --clean` / `apk --clean` / `apk --force`, or should we not try and be incremental for some of the commands?

We should fix the travis issues and the formatting nits, then lgtm.

Hixie commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -179,227 +191,342 @@ class ApkCommand extends FlutterCommand {
+  String keyPassword;
+
+  if (keystoreInfo == null) {
+    printError('Signing the APK using the debug keystore');
```
nit: put a period at the end of this line for consistency with the other output

mpcomplete commented:
Re your comments:

> when there's no building required, start could just say Starting foo on bar... - we don't need to say APK up to date. Skipping build step.

OK, I made this a printTrace so we can at least see it with -v.

> We should still say something about skipping the build if the user has explicitly asked for a build like via flutter apk (or perhaps never do timestamp checks for explicit builds?)

I made it so `flutter apk` always forces a build. Only `flutter start` does a dep check.

> agree w/ that long pause at the start, that we should say something. For things we know will take a while, perhaps also add the duration at the end. So Building APK..., and then APK built to build/app.apk (5.3s).

OK, added a message. I skipped the timestamp, because those are visible when you use -v.

> what can the user currently do if they know that the app needs to be re-built, but our system thinks it's up-to-date? Is there a build --clean / apk --clean / apk --force, or should we not try and be incremental for some of the commands?

Yeah, right now `flutter apk` always forces a rebuild.

mpcomplete committed:
'flutter start' calls 'flutter apk' if necessary.

flutter start no longer depends on a pre-built SkyShell.apk. It builds a
new one, as long as an AndroidManifest.xml exists.

We rebuild the .apk every time either AndroidManifest.xml or
flutter.yaml changes.
